source/MRMesh/MRVoxelsConversionsByParts.cpp

```diff
@@ -0,0 +1,204 @@
+#include "MRVoxelsConversionsByParts.h"
+#include "MREdgePaths.h"
+#include "MRMesh.h"
+#include "MRMeshTrimWithPlane.h"
+#include "MRPlane3.h"
+#include "MRStringConvert.h"
+#include "MRTimer.h"
+#include "MRVDBConversions.h"
+
+#include <fmt/format.h>
+
+namespace
+{
+
+using namespace MR;
+
+void clearPartMapping( PartMapping& mapping )
+{
+    if ( mapping.src2tgtFaces )
+        mapping.src2tgtFaces->clear();
+    if ( mapping.src2tgtVerts )
+        mapping.src2tgtVerts->clear();
+    if ( mapping.src2tgtEdges )
+        mapping.src2tgtEdges->clear();
+    if ( mapping.tgt2srcFaces )
+        mapping.tgt2srcFaces->clear();
+    if ( mapping.tgt2srcVerts )
+        mapping.tgt2srcVerts->clear();
+    if ( mapping.tgt2srcEdges )
+        mapping.tgt2srcEdges->clear();
+}
+
+void sortEdgePaths( const Mesh& mesh, std::vector<EdgePath>& paths )
+{
+    std::sort( paths.begin(), paths.end(), [&] ( const EdgePath& ep1, const EdgePath& ep2 )
+    {
+        auto org1 = mesh.orgPnt( ep1.front() );
+        auto org2 = mesh.orgPnt( ep2.front() );
+        auto dest1 = mesh.destPnt( ep1.front() );
+        auto dest2 = mesh.destPnt( ep2.front() );
+        if ( org1 != org2 )
+            return std::tie( org1.y, org1.z ) < std::tie( org2.y, org2.z );
+        else
+            return std::tie( dest1.y, dest1.z ) < std::tie( dest2.y, dest2.z );
+    } );
+}
+
+} // namespace
+
+namespace MR
+{
+
+tl::expected<void, std::string>
+mergeGridPart( Mesh &mesh, std::vector<EdgePath> &cutContours, FloatGrid &&grid, const Vector3f &voxelSize,
+               float leftCutPosition, float rightCutPosition, const MergeGridPartSettings &settings )
+{
+    MR_TIMER
+
+    Timer timer( "convert grid to mesh" );
+    auto res = gridToMesh( std::move( grid ), GridToMeshSettings {
+            .voxelSize = voxelSize,
+    } );
+    if ( !res.has_value() )
+        return tl::make_unexpected( res.error() );
+    auto part = std::move( *res );
+
+    if ( settings.preCut )
+    {
+        timer.restart( "pre-cut callback" );
```

Comment:
i think timers for callbacks shoud be inside them (user will place it, no need to do it for him)

Reply:
Fixed.